http-webhook.md

```diff
@@ -0,0 +1,333 @@
+Cache-Control header with the "private" option.
+
+Because of the security weaknesses associated with the URI method (see
+[RFC6750, Section 5][RFC6750]), including the high likelihood that the URL
+containing the access token will be logged, it SHOULD NOT be used unless it is
+impossible to transport the access token in the "Authorization" request header
+field or the HTTP request entity-body. All further caveats cited in
+[RFC6750][RFC6750] apply equivalently.
+
+## 4. Abuse Protection
+
+Any system that allows registration of and delivery of notifications to
+arbitary HTTP endpoints can potentially be abused such that someone maliciously
+or inadvertently registers the address of a system that does not expect such
+requests and for which the registering party is not authorized to perform such
+a registration. In extreme cases, a notification infrastructure could be abused
+to launch denial-of-service attacks against an arbitrary web-site.
+
+To protect the sender from being abused in such a way, a legitimate delivery
+target needs to indicate that it agrees with notifications being delivered to it.
+
+Reaching the delivery agreement is realized using the following validation
+handshake. The handshake can either be executed immediately at registration
+time or as a "pre-flight" request immediately preceding a delivery. If the
+handshake outcome is that the delivery is denied, the sender MUST NOT deliver
+events to the target.
+
+It is important to understand that the handshake does not aim to establish
+an authentication or authorization context. It only serves to protect the
+sender from being told to a push to a destination that is not expecting the
+traffic. While this specification mandates use of an authorization model, this
+mandate is not sufficient to protect any arbitrary website from unwanted
+traffic if that website doesn't implement access control and therefore ignores
+the `Authorization` header.
+
+Delivery targets SHOULD support the abuse protection feature. If a target does
+not support the feature, the sender MAY choose not to send to the target, at
+all, or send only at a very low request rate.
+
+### 4.1.1. Validation request
+
+The validation request uses the HTTP [OPTIONS][OPTIONS] method. The request
+is directed to the exact resource target URI that is being registered.
+
+With the validation request, the sender asks the target for permission to
+send notifications, and it can declare a desired request rate (requests per
+minute).
+
+The delivery target will respond with a permission statement and the permitted
+request rate.
+
+The following header fields MUST be included in the validation request:
+
+#### 4.1.1.2. WebHook-Request-Origin
+
+The `WebHook-Request-Origin` header requests permission to send notifications
+from this sender, and contains a DNS expression that identifies the sending
+system, for example "eventemitter.example.com". The value is meant to summarily
```

Comment:
Who/what guarantees this identity really belongs to the request sender?

Would it make sense to build on top of SPIFFE/Spire/Istio?

Reply:
I am trying to [explain here](https://github.com/cloudevents/spec/pull/155/files#diff-c47b1661017c688fecb70618769650f1R183) that Origin is not about identity. Authorization takes care of establishing identity where needed.

There are three roles here: (a) sender, (b) target, and (c) the 3rd party that tells the sender to send to the target. The threat vector here is that (c) tells (a) to send to (b) without (b) being ok to receive events from (a). 

If (b) is an plain website running on nginx without authentication turned on, it's not going to benefit from any fancy identity model. What protects it from unwanted push traffic is that it simply doesn't respond to the abuse handshake with a positive acknowledgement. "Origin" allows a site to differentiate between different senders within the scope of this mechanism. Since the mechanism is designed to mitigate the threat explained above, Origin doesn't need further protection, because (c) can't manipulate it. 

If I am operating (a) as "push.fabrikam.com" and were to steal someone else's identity, say "push.contoso.com" for the origin, because (b) has that name whitelisted but not mine, I'm not gaining any better access because I still need proper credentials. But I just allowed (c) to abuse my service to bombard (b) with invalid requests and therefore risk blacklisting of my originating IP address. So making a false statement about "Origin" is actually self-defeating with what this mechanism is aiming at. 

If `Origin` needs to be further authenticated, that should be done using the existing authNZ model, e.g. you could use TLS client auth to establish the identity, or you could include the Origin value into the bearer token.

Reply:
I'm a bit torn here because this specifies a perfectly valid model and I like the idea of dynamic rate limiting. A few questions though:

1. Why not actually reuse the CORS headers here?
2. Many clouds implement this by requiring (c) to prove ownership of (b) before allowing (a) to point to (c). Should such a cloud have a _separate_ auth mechanism just for cloud events?